src/services/createRouter.ts

```diff
@@ -123,7 +126,18 @@ export function createRouter<const TRoutes extends Routes, const TOptions extend
         throw new Error(`Switch is not exhaustive for before hook response status: ${JSON.stringify(beforeResponse satisfies never)}`)
     }
 
-    propStore.setProps(to)
+    try {
+      propStore.setProps(to)
+    } catch (error) {
+      if (error instanceof RouterPushError) {
+        await push(...error.to as Parameters<RouterPush>)
+        return
+      }
+
+      if (error instanceof RouterRejectionError) {
+        setRejection(error.type)
+      }
+    }
```

Comment:
I don't love having all this logic inserted in an already big file and a big function. Open to suggestions